### PR TITLE
Harden asn1 oid config loader to invalid inputs

### PR DESCRIPTION
**Pull request by nhorman**
Harden ans1 oid loader to invalid inputs
    
    In the event that a config file contains this sequence:
    =======
    openssl_conf = openssl_init
    
    config_diagnostics = 1
    
    [openssl_init]
    oid_section = oids
    
    [oids]
    testoid1 = 1.2.3.4.1
    testoid2 = A Very Long OID Name, 1.2.3.4.2
    testoid3 = ,1.2.3.4.3
    ======
    
    The leading comma in testoid3 can cause a heap buffer overflow, as the
    parsing code will move the string pointer back 1 character, thereby
    pointing to an invalid memory space
    
    correct the parser to detect this condition and handle it by treating it
    as if the comma doesn't exist (i.e. an empty long oid name)

##### Checklist
- [x] tests are added or updated

### CODE REVIEW

**Comment by tmshort** [comment]
I would prefer that `asn1test.cnf` be renamed to `test_asn1_parse.cnf` or similar, to make it clear that it's just for the *test_asn1_parse* test. As it stands `asn1test.cnf` is a bit too generic a name (i.e. does it apply to all ASN.1 tests?).

**Comment by jamuir** [comment]
There are two separate commits in this PR.

Are they together b/c the new test covers both of the issues described in the commit messages?

**Comment by nhorman** [comment]
The only reason they are together is because while I was using asn1parse to validate that the config is parsed properly, I found that the genrsa option was broken, so I fixed them together.  Just me being a bit lazy

**Review by jamuir — APPROVED** [review]
looks good.

I think Todd's comment about renaming the config file makes sense.

When this is merged, I think the commits should be kept separate (i.e. do not squash).

File: apps/asn1parse.c [hunk 1]
```diff
@@ -178,7 +178,7 @@ int asn1parse_main(int argc, char **argv)
 
     if ((buf = BUF_MEM_new()) == NULL)
         goto end;
-    if (informat == FORMAT_PEM) {
+    if (genstr == NULL && informat == FORMAT_PEM) {
```
Comment (t8m) [inline apps/asn1parse.c:181]: Not just genstr but also the genconf option is broken in the same way.

**Comment by nhorman** [comment]
@jamuir ack, the intent was to keep them separate.

File: apps/asn1parse.c (hunk 1 above)
Comment (nhorman) [reply apps/asn1parse.c:181]: Yup, I'll fix that up

**Review by t8m — APPROVED** [review]

**Comment by drewwells** [comment]
I'm not an expert on how to install this in an alternate location on linux, I'll give it another go today.

**Review by hlandau — APPROVED** [review]

**Comment by nhorman** [comment]
merged to master, needs separate branches for 3.0, 3.1 and 3.2

**Comment by t8m** [comment]
I was able to cherry-pick this to 3.2 without any changes.

**Pull request closed without merging**